### ISSUE TITLE
Why make all the device generate from the same latent (noise)?

### ISSUE THREAD

**Issue by mchong6** [Tencent-Hunyuan/SRPO#39]
1)

 #### it is important to make all device generate with the same latent
    dist.broadcast(input_latents, src=0)

I am not sure why this is done. If I understand correctly, each device will get a different prompt, but they are starting from the same random noise? Why is this necessary?

2) Also from what I can see, instead of optimizing over all timesteps, you are only optimizing the middle timesteps?

3) 
            latents = noisy.detach()
            transformer.train()
            latents = latents.requires_grad_(True)

Why do you need to enable grad for latents? latents is not being optimized.

**Comment by xwwshen** [comment]
> #### it is important to make all device generate with the same latent
> 
> ```
> dist.broadcast(input_latents, src=0)
> ```
> 
> I am not sure why this is done. If I understand correctly, each device will get a different prompt, but they are starting from the same random noise? Why is this necessary?
> 
> 2. Also from what I can see, instead of optimizing over all timesteps, you are only optimizing the middle timesteps?
> 3. ```
>         latents = noisy.detach()
>         transformer.train()
>         latents = latents.requires_grad_(True)
>    ```
> 
> Why do you need to enable grad for latents? latents is not being optimized.

Thanks for your questions：

Yes — each device receives a different prompt but shares the same random seed/initial noise, so the Direct‑align noise is identical across devices. Because we directly use analytical gradients from a reward model, that signal can be noisy and imperfect for optimization; synchronizing per‑device randomness empirically yields cleaner, more stable outputs and reduces variance across replicas.

**On optimizing only the middle timesteps**: empirically, very early timesteps are unstable and prone to collapse, so we avoid the extreme early region.  late timesteps tend to pick up high‑frequency biases of the reward model and can be “hacked” (like overfit to special color). Focusing on middle and early timesteps gives a more stable, meaningful learning signal.

**Why enable grad for latents?** The preceding operations may produce no‑grad tensors. Calling latents.requires_grad_(True) ensures gradients can flow back through the latents (and therefore into the model), making backpropagation possible.
